market_maker_keeper/ddex_market_maker_keeper.py

```diff
@@ -0,0 +1,240 @@
+                                                                              request_kwargs={"timeout": self.arguments.rpc_timeout}))
+        self.web3.eth.defaultAccount = self.arguments.eth_from
+        self.our_address = Address(self.arguments.eth_from)
+
+        self.pair = self.arguments.pair.upper()
+        self.token_buy = ERC20Token(web3=self.web3, address=Address(self.arguments.buy_token_address))
+        self.token_sell = ERC20Token(web3=self.web3, address=Address(self.arguments.sell_token_address))
+        self.min_eth_balance = Wad.from_number(self.arguments.min_eth_balance)
+        self.bands_config = ReloadableConfig(self.arguments.config)
+        self.price_max_decimals = None
+        self.amount_max_decimals = None
+        self.gas_price = GasPriceFactory().create_gas_price(self.arguments)
+        self.price_feed = PriceFeedFactory().create_price_feed(self.arguments)
+        self.spread_feed = create_spread_feed(self.arguments)
+        self.order_history_reporter = create_order_history_reporter(self.arguments)
+
+        self.history = History()
+        self.zrx_exchange = ZrxExchange(web3=self.web3, address=Address(self.arguments.exchange_address))
+        self.ddex_api = DdexApi(self.web3,
+                                self.arguments.ddex_api_server,
+                                self.arguments.ddex_api_timeout)
+
+        self.order_book_manager = OrderBookManager(refresh_frequency=self.arguments.refresh_frequency, max_workers=1)
+        self.order_book_manager.get_orders_with(lambda: self.ddex_api.get_orders(self.pair))
+        self.order_book_manager.cancel_orders_with(lambda order: self.ddex_api.cancel_order(order.order_id))
+        self.order_book_manager.enable_history_reporting(self.order_history_reporter, self.our_buy_orders, self.our_sell_orders)
+        self.order_book_manager.start()
+
+    def main(self):
+        with Lifecycle(self.web3) as lifecycle:
+            lifecycle.initial_delay(10)
+            lifecycle.on_startup(self.startup)
+            lifecycle.every(1, self.synchronize_orders)
+            lifecycle.on_shutdown(self.shutdown)
+
+    def startup(self):
+        self.approve()
+
+        # Get maximum number of decimals for prices and amounts.
+        # Ddex API enforces it.
+        markets = self.ddex_api.get_markets()['data']['markets']
+        market = next(filter(lambda item: item['id'] == self.pair, markets))
+
+        self.price_max_decimals = market['pricePrecision']
+        self.amount_max_decimals = market['amountDecimals']
+        #TODO minOrderSize ?
+
+    def shutdown(self):
+        self.order_book_manager.cancel_all_orders()
+
+    def approve(self):
+        self.zrx_exchange.approve([self.token_sell, self.token_buy], directly(gas_price=self.gas_price))
+
+    def our_total_balance(self, token: ERC20Token) -> Wad:
+        return token.balance_of(self.our_address) #TODO use ddex_api.get_balances ?
```

Comment:
I don't think any of these approaches is significantly better than the other. In case of _Paradex_ we use ERC20 token balance and subtract the amounts locked in open orders, exactly as you've done it here.

I am open to discussion though if you want to do it the other way around.